[TableGen] Allow empty terminator in SequenceToOffsetTable

s-barannikov opened the pull request:
Some clients do not want to emit a terminator after each sub-sequence (they have other means of determining the length of sub-sequences).

This moves `Term` argument from `emit` method to the constructor and makes it optional. It couldn't be made optional while still on the `emit` method because if the terminator wasn't specified, it has to be taken into account in `layout` method as well.

The fact that `layout` method was called is now recorded in a dedicated member variable, `IsLaidOut`. `Entries != 0` can no longer be used to reliably check if `layout` method was called because it may be zero for a different reason: the terminator wasn't specified and all added sequences (if any) were empty.

This reduces the size of `*LaneMaskLists` and `*SubRegIdxLists` a bit and resolves the removed TODO.

s-barannikov committed:
[TableGen] Allow empty terminator in SequenceToOffsetTable

Some clients do not want to emit a terminator after each sub-sequence
(they have other means of determining the length of the sub-sequence).

This moves `Term` argument from `emit` method to the constructor
and makes it optional. It couldn't be made optional while still
on the `emit` method because if the terminator wasn't specified,
it has to be taken into account in `layout` method as well.

The fact that `layout` method was called is now recorded in a dedicated
member variable, `IsLaidOut`. `Entries != 0` can no longer be used
to reliably check if `layout` method was called because it may be zero
for a different reason: the terminator wasn't specified and all added
sequences (if any) were empty.

This reduces the size of `*LaneMaskLists` and `*SubRegIdxLists` a bit
and resolves the removed FIXME.

jurahul commented on llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:
```diff
@@ -97,25 +103,27 @@ class SequenceToOffsetTable {
       I->second = Entries;
       // Include space for a terminator.
-      Entries += I->first.size() + 1;
+      Entries += I->first.size() + Terminator.has_value();
```
nit: Can we just do:

```
if (Terminator.has_value())
  Entries += Seqs.size();
```

after the loop?

s-barannikov replied:
Done

s-barannikov replied:
Unfortunately, that didn't work out. `I->second` above needs the adjusted value.

jurahul replied:
I see that, thanks for trying it out.

s-barannikov committed:
Simplify the loop in `layout`

jurahul commented on llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:
```diff
@@ -127,10 +135,10 @@ class SequenceToOffsetTable {
     if (!EmitLongStrLiterals) {
       OS << Decl << " = {\n";
-      emit(OS, printChar, "0");
+      emit(OS, printChar);
```
Do we need to print the terminator here?

s-barannikov replied:
It is printed by `emit` after each subsequence, assuming the terminator was provided.
The line below adds an extra '0' that I believe it should not do (and I think I ran into this before), but this is a pre-existing behavior and may be related to the issue I'm observing (see my comment below).

s-barannikov replied:
The extra '\0' printing was added in https://reviews.llvm.org/D126136 and looks intentional. This way the generated data is the same regardless of EmitLongStrLiterals. If the option is enabled, the whole data is printed as a single string literal, and this string literal has an implicit terminating '\0' according to C. That change made the other case emit this extra '\0'.

It doesn't appear related to the issue with zero-sized arrays.

s-barannikov commented:
There is a bunch of warnings I need to deal with: `warning: zero size arrays are an extension [-Wzero-length-array]`
that turn into an error when building with MSVC: `error C2466: cannot allocate an array of constant size 0`.

s-barannikov committed:
Print a dummy element if the array would be empty otherwise

s-barannikov commented:
> There is a bunch of warnings I need to deal with: `warning: zero size arrays are an extension [-Wzero-length-array]` that turn into an error when building with MSVC: `error C2466: cannot allocate an array of constant size 0`.
> 
> ```
>     84  extern const uint16_t XCoreSubRegIdxLists[] = {
>     85    /* 0 */ 
>     86  };
> ```

I couldn't think of anything better than to print a dummy default-constructed element if the array would otherwise be empty.

s-barannikov committed:
Revert "Simplify the loop in `layout`"

This reverts commit a3676eda47a65ca50380c717df149bb178779e85.

s-barannikov committed:
Provide a default value for the parameter of the constructor

Most of the clients use the default value for the terminator.
This reduces change reduces the diff.

s-barannikov committed:
Fix failing test
